docs/man/man3/Learner.3
.TH "Learner" 3 "Fri May 14 2021" "Version 2.1.0" "Text-Classifier" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Learner \- This class describes a learner\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <learner\&.hpp>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBLearner\fP (void)"
.br
.RI "Constructs a new instance\&. "
.ti -1c
.RI "\fBLearner\fP (char *argv[], int &argc)"
.br
.RI "Constructs a new instance\&. "
.ti -1c
.RI "\fB~Learner\fP (void)"
.br
.RI "Destroys the object\&. "
.ti -1c
.RI "std::vector< \fBVocabulary\fP > \fBget_Learners\fP (void) const"
.br
.RI "Gets the learners\&. "
.ti -1c
.RI "std::vector< std::string > \fBget_InputCorpusFiles\fP (void) const"
.br
.RI "Gets the input corpus files\&. "
.ti -1c
.RI "std::string \fBget_VocabularyFile\fP (void) const"
.br
.RI "Gets the vocabulary file\&. "
.ti -1c
.RI "void \fBset_Learners\fP (std::vector< \fBVocabulary\fP > newLearners)"
.br
.RI "Sets the learners\&. "
.ti -1c
.RI "void \fBset_InputCorpusFiles\fP (std::vector< std::string > newInputCorpusFiles)"
.br
.RI "Sets the input corpus files\&. "
.ti -1c
.RI "void \fBset_VocabularyFile\fP (std::string newVocabularyFile)"
.br
.RI "Sets the vocabulary file\&. "
.ti -1c
.RI "void \fBlearnAndStoreData\fP (void)"
.br
.RI "Generates the proabilities of all the tokens in vocabulary and stores them in a file called 'aprendizaje_DATATYPE\&.txt'\&. The file format is: Number of corpus documents: <integer> Number of words in corpus: <integer> Word:<string> Frec:<int> LogProb:<float> Word:<string> Frec:<int> LogProb:<float> Word:<string> Frec:<int> LogProb:<float> "
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "std::vector< \fBVocabulary\fP > \fBlearners_\fP"
.br
.ti -1c
.RI "std::vector< std::string > \fBinputCorpusFiles_\fP"
.br
.ti -1c
.RI "std::string \fBvocabularyFile_\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
This class describes a learner\&. 
.PP
Definition at line 37 of file learner\&.hpp\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Learner::Learner (char * argv[], int & argc)"

.PP
Constructs a new instance\&. 
.PP
\fBParameters\fP
.RS 4
\fIargv\fP The arguments array 
.br
\fIargc\fP The count of arguments 
.RE
.PP

.PP
Definition at line 41 of file learner\&.cpp\&.
.PP
.nf
41                                          {
42     // Vocabulary file must be in 3rd place as shown in manual
43     set_VocabularyFile(argv[2]);
44     for (int i = 3; i < argc; i++) {
45         Chrono chrono;
46         chrono\&.startChrono();
47         std::string tmp = "\&.\&./outputs/corpus_";
48         tmp += argv[i];
49         tmp += "\&.txt";
50         Vocabulary newVoc;
51         newVoc\&.readVocabulary(vocabularyFile_);
52         newVoc\&.generateVocabulary(tmp, true);
53         learners_\&.push_back(newVoc);
54         inputCorpusFiles_\&.push_back(argv[i]);
55         chrono\&.stopChrono();
56         std::cout << std::endl << "Elapsed generating tokens and vocabulary time: " << chrono\&.get_Seconds(5) << " seconds for " << argv[i] << " data type\&." << std::endl;
57     }
58     learnAndStoreData();
59 }
.fi
.SH "Member Function Documentation"
.PP 
.SS "std::vector< std::string > Learner::get_InputCorpusFiles (void) const"

.PP
Gets the input corpus files\&. 
.PP
\fBReturns\fP
.RS 4
The input corpus files\&. 
.RE
.PP

.PP
Definition at line 81 of file learner\&.cpp\&.
.PP
.nf
81                                                               {
82     return inputCorpusFiles_;
83 }
.fi
.SS "std::vector< \fBVocabulary\fP > Learner::get_Learners (void) const"

.PP
Gets the learners\&. 
.PP
\fBReturns\fP
.RS 4
The learners\&. 
.RE
.PP

.PP
Definition at line 72 of file learner\&.cpp\&.
.PP
.nf
72                                                        {
73     return learners_;
74 }
.fi
.SS "std::string Learner::get_VocabularyFile (void) const"

.PP
Gets the vocabulary file\&. 
.PP
\fBReturns\fP
.RS 4
The vocabulary file\&. 
.RE
.PP

.PP
Definition at line 90 of file learner\&.cpp\&.
.PP
.nf
90                                                  {
91     return vocabularyFile_;
92 }
.fi
.SS "void Learner::set_InputCorpusFiles (std::vector< std::string > newInputCorpusFiles)"

.PP
Sets the input corpus files\&. 
.PP
\fBParameters\fP
.RS 4
\fInewInputCorpusFiles\fP The new input corpus files 
.RE
.PP

.PP
Definition at line 108 of file learner\&.cpp\&.
.PP
.nf
108                                                                             {
109     inputCorpusFiles_ = newInputCorpusFiles;
110 }
.fi
.SS "void Learner::set_Learners (std::vector< \fBVocabulary\fP > newLearners)"

.PP
Sets the learners\&. 
.PP
\fBParameters\fP
.RS 4
\fInewLearners\fP The new learners 
.RE
.PP

.PP
Definition at line 99 of file learner\&.cpp\&.
.PP
.nf
99                                                              {
100     learners_ = newLearners;
101 }
.fi
.SS "void Learner::set_VocabularyFile (std::string newVocabularyFile)"

.PP
Sets the vocabulary file\&. 
.PP
\fBParameters\fP
.RS 4
\fInewVocabularyFile\fP The new vocabulary file 
.RE
.PP

.PP
Definition at line 117 of file learner\&.cpp\&.
.PP
.nf
117                                                              {
118     vocabularyFile_ = newVocabularyFile;
119 }
.fi


.SH "Author"
.PP 
Generated automatically by Doxygen for Text-Classifier from the source code\&.


src/learner.cpp
/*=======================================================================================
=========================================================================================
	=                                                                              =
	=            Proyect:       Text Classifier                                    =
	=            File name:     learner.cpp                                        =
	=            Author:        Adrián Epifanio Rodríguez Hernández                =
	=            Date:          21/04/2021                                         =
	=            Subject:       Advanced Artificial Inteligence                    =
	=            Language:      C++                                                =
	=            Email:         alu0101158280@ull.edu.es                           =
	=            Place:         Universidad De La Laguna                           =
	=                           Escuela Superior de Ingeniería y Tecnología        =
	=                                                                              =
=========================================================================================
=======================================================================================*/
/*
* @Author: Adrian Epifanio
* @Date:   2021-05-01 12:27:32
* @Last Modified by:   Adrian Epifanio
* @Last Modified time: 2021-05-09 16:50:24
*/
/*------------------  FUNCTIONS  -----------------*/

#include "../include/learner.hpp"

/*------------------------------------------------*/

/**
 * @brief      Constructs a new instance.
 */
Learner::Learner (void) {
	set_VocabularyFile("");
}

/**
 * @brief      Constructs a new instance.
 *
 * @param      argv  The arguments array
 * @param      argc  The count of arguments
 */
Learner::Learner (char* argv[], int& argc) {
	// Vocabulary file must be in 3rd place as shown in manual
	set_VocabularyFile(argv[2]);
	for (int i = 3; i < argc; i++) {
		Chrono chrono;
		chrono.startChrono();
		std::string tmp = "../outputs/corpus_";
		tmp += argv[i];
		tmp += ".txt";
		Vocabulary newVoc;
		newVoc.readVocabulary(vocabularyFile_);
		newVoc.generateVocabulary(tmp, true);
		learners_.push_back(newVoc);
		inputCorpusFiles_.push_back(argv[i]);
		chrono.stopChrono();
		std::cout << std::endl << "Elapsed generating tokens and vocabulary time: " << chrono.get_Seconds(5) << " seconds for " << argv[i] << " data type." << std::endl;
	}
	learnAndStoreData();
}

/**
 * @brief      Destroys the object.
 */
Learner::~Learner (void) {
}

/**
 * @brief      Gets the learners.
 *
 * @return     The learners.
 */
std::vector<Vocabulary> Learner::get_Learners (void) const {
	return learners_;
}

/**
 * @brief      Gets the input corpus files.
 *
 * @return     The input corpus files.
 */
std::vector<std::string> Learner::get_InputCorpusFiles (void) const {
	return inputCorpusFiles_;
}

/**
 * @brief      Gets the vocabulary file.
 *
 * @return     The vocabulary file.
 */
std::string Learner::get_VocabularyFile (void) const {
	return vocabularyFile_;
}

/**
 * @brief      Sets the learners.
 *
 * @param[in]  newLearners  The new learners
 */
void Learner::set_Learners (std::vector<Vocabulary> newLearners) {
	learners_ = newLearners;
}

/**
 * @brief      Sets the input corpus files.
 *
 * @param[in]  newInputCorpusFiles  The new input corpus files
 */
void Learner::set_InputCorpusFiles (std::vector<std::string> newInputCorpusFiles) {
	inputCorpusFiles_ = newInputCorpusFiles;
}

/**
 * @brief      Sets the vocabulary file.
 *
 * @param[in]  newVocabularyFile  The new vocabulary file
 */
void Learner::set_VocabularyFile (std::string newVocabularyFile) {
	vocabularyFile_ = newVocabularyFile;
}

/**
 * @brief      Generates the proabilities of all the tokens in vocabulary and
 *             stores them in a file called "aprendizaje_DATATYPE.txt".
 *             The file format is:
 *             Number of corpus documents: <integer>
 *             Number of words in corpus: <integer>
 *             Word:<string> Frec:<int> LogProb:<float>
 *             Word:<string> Frec:<int> LogProb:<float>
 *             Word:<string> Frec:<int> LogProb:<float>
 */
void Learner::learnAndStoreData (void) {
	for (unsigned i = 0; i < learners_.size(); i++) {
		Chrono chrono;
		chrono.startChrono();
		std::string fileName = "../outputs/aprendizaje";
		fileName += inputCorpusFiles_[i][0];
		fileName += ".txt";
		//std::string fileName = "../outputs/aprendizaje_" + inputCorpusFiles_[i] + ".txt";
		std::fstream file(fileName, std::ios::out);
		if (file.fail()) {
			std::cout << "Error while storing learned data \"" << fileName << "\" is not valid document" << std::endl;
			exit(1);
		} 
		else {
			int tokenAmmount = learners_[i].get_NTokens();
			int vocSize = learners_[i].get_VocabularyCounter();
			std::string data = "Numero de documentos del corpus: " + std::to_string(learners_[i].get_NLines());
			data += "\nNumero de palabras del corpus: " + std::to_string(tokenAmmount);
			for (auto tmp : learners_[i].get_Vocabulary()) {
				std::string line =  "\nPalabra: " + tmp.get_Name();
				// Tabs for columns
				/*
				while (line.length() < 50) {
					line += " ";
				}
				data += line + "\t\tFrec: " + std::to_string(tmp.get_Ammount()) + "  \t\tLogProb: " + std::to_string(tmp.generateLogProb(vocSize, tokenAmmount));
				*/
				// Without tabs
				data += line + " Frec: " + std::to_string(tmp.get_Ammount()) + " LogProb: " + std::to_string(tmp.generateLogProb(vocSize, tokenAmmount));
			}
			file << data;
			file.close();
		}
		chrono.stopChrono();
		std::cout << std::endl << "Elapsed time for calculating probabilities: " << chrono.get_Seconds(5) << " seconds for " << inputCorpusFiles_[i] << " data type." << std::endl;
	}
}
